### ISSUE TITLE
New TP server table suggestions

### ISSUE THREAD

**Issue by traeak** [apache/trafficcontrol#4824]
## I'm submitting a ...

- [ ] bug report
- [ ] new feature / enhancement request
- [x] improvement request (usability, performance, tech debt, etc.)
- [ ] other 

## Traffic Control components affected ...

- [ ] CDN in a Box
- [ ] Documentation
- [ ] Grove
- [ ] Traffic Control Client
- [ ] Traffic Monitor
- [ ] Traffic Ops
- [ ] Traffic Ops ORT
- [x] Traffic Portal
- [ ] Traffic Router
- [ ] Traffic Stats
- [ ] Traffic Vault
- [ ] unknown

## Current behavior:

Just some nits on the new TP servers portal.  Purely my opinion.

- How to search isn't initially obvious.  Per column search is a good feature though.
- Single line table results initially have terrible elision as everything looks the same.
- It would be *lovely* to have tooltip hover for non column expanded items.

**Comment by mitchell852** [comment]
> * How to search isn't initially obvious.  Per column search is a good feature though.

Per column table search makes searching very powerful but i agree, it isn't very obvious how to do it. i also think a global, across all column search (like the old table had) is still very valuable to the non-power user.

**Comment by rawlinp** [comment]
I'll add on that I don't really mind adjusting column widths as long as TP will remember the widths, but it would be nice if adjusting one column's width didn't eat into the width of the column next to it. That would make adjusting the column widths much easier.

**Comment by ocket8888** [comment]
> _"Single line table results initially have terrible elision as everything looks the same."_

> _"In linguistics, an elision or deletion is the omission of one or more sounds in a word or phrase. The word elision is frequently used in linguistic description of living languages, and deletion is often used in historical linguistics for a historical sound change." - Wikipedia article on elision_

I'm not sure how to apply that in this context

> _"It would be lovely to have tooltip hover for non column expanded items."_

What are "non column expanded items"? You want a tooltip on hover that shows all of the columns that are hidden? Or a tooltip that shows cell content when the column is so thin that content gets cut off?

**Comment by mitchell852** [comment]
there's also an issue where if you apply a filter to a column and then hide that column, the filter remains in place. if the column, is hidden the filter should be removed.

**Comment by traeak** [comment]
No a tooltip over the items in a column that have been truncated.  It's possible (but not necessary) to more intelligently shorten what's being displayed.  Fixing that is more of an excuse to write something that uses dynamic programming than anything else.

**Comment by mitchell852** [comment]
> No a tooltip over the items in a column that have been truncated.

![image](https://user-images.githubusercontent.com/251272/85728996-a6ae5480-b6b5-11ea-8a5e-5bf08d50108c.png)

yeah, it would be nice to mouseover each cell and see the full text

**Comment by ocket8888** [comment]
That'll mean the current tooltip for "Status" which shows the offline reason for offline servers - which was present in the old table - will need to go away. Is that alright?

**Comment by mitchell852** [comment]
> That'll mean the current tooltip for "Status" which shows the offline reason for offline servers - which was present in the old table - will need to go away. Is that alright?

I think that's a fair trade off

**Comment by mitchell852** [comment]
![image](https://user-images.githubusercontent.com/251272/86023661-2e170300-b9e9-11ea-8d58-ad6efbb13f03.png)

clicking on an empty cell doesn't trigger row click.

**Comment by ocket8888** [comment]
hm. It ought to. That's a bug, tbh, not a suggestion.

**Comment by mitchell852** [comment]
> hm. It ought to. That's a bug, tbh, not a suggestion.

i'll create a bug

**Comment by dg4prez** [comment]
New table looks great but I'm already missing the number of results per page - can we add that back in? It's nice to have all when you want it.

**Comment by mitchell852** [comment]
also, a clear all filters button would be handy...

**Comment by mitchell852** [comment]
another one: pagination is not persisted. i.e. if you go to page 5, click a server then return to the grid, you are back on page 1.

**Comment by ocket8888** [comment]
Here's my suggestion: the new search added in #4891 could be made [fuzzy](https://en.wikipedia.org/wiki/Approximate_string_matching).

**Comment by mitchell852** [comment]
> Here's my suggestion: the new search added in #4891 could be made [fuzzy](https://en.wikipedia.org/wiki/Approximate_string_matching).

can it? i'll have to look into ag-grid's `setQuickFilter` api to see if it supports anything like that.

**Comment by mitchell852** [comment]
> Here's my suggestion: the new search added in #4891 could be made [fuzzy](https://en.wikipedia.org/wiki/Approximate_string_matching).

@ocket8888 can u make a separate issue for that?

**Comment by mitchell852** [comment]
> I'll add on that I don't really mind adjusting column widths as long as TP will remember the widths, but it would be nice if adjusting one column's width didn't eat into the width of the column next to it. That would make adjusting the column widths much easier.

@rawlinp - TP should remember the widths you set but the eating into the width part might be just the way the grid behaves. Not sure there's a reasonable fix for that.

**Comment by mitchell852** [comment]
> another one: pagination is not persisted. i.e. if you go to page 5, click a server then return to the grid, you are back on page 1.

issue created - #5384

**Comment by ocket8888** [comment]
I don't really know what re-sizing columns would look like if it didn't "eat into" the width of other columns. Unless you'd want the table to have a horizontal scrollbar if the size got bigger than the viewport?

**Comment by rawlinp** [comment]
It would be a better experience if when I widen column A it didn't also narrow column B at the same time. It means needing to manually resize _all_ columns whenever you want to resize a single column.

**Comment by mitchell852** [comment]
> It would be a better experience if when I widen column A it didn't also narrow column B at the same time. It means needing to manually resize _all_ columns whenever you want to resize a single column.

yeah, i agree. it's not much fun. FWIW you can double click the column separator to expand it to fit the column contents...

**Comment by ocket8888** [comment]
Right, but if you widen column A without narrowing column B then where does the space come from? What are you expanding into?

**Comment by rawlinp** [comment]
>  FWIW you can double click the column separator to expand it to fit the column contents...

Oh, that's a neat trick I would've never figured out on my own!

> Right, but if you widen column A without narrowing column B then where does the space come from? What are you expanding into?

_The Void_. It should expand horizontally -- it already allows you to scroll horizontally if the size is bigger than the viewport.

**Comment by ocket8888** [comment]
Okay. I kind of hate horizontal scrollbars, so the current implementation is built with avoiding them if possible in mind, but it's totally possible to bring them back - that's actually the default behavior.

**Comment by rawlinp** [comment]
I don't actually see a horizontal scrollbar, but I am able to scroll horizontally.
